src/table/src/metadata.rs

```diff
@@ -231,33 +239,54 @@ impl TableMeta {
     ) -> Result<TableMetaBuilder> {
         let table_schema = &self.schema;
         let mut meta_builder = self.new_meta_builder();
-
-        // Check whether columns to add are already existing.
-        for request in requests {
-            let column_name = &request.column_schema.name;
-            ensure!(
-                table_schema.column_schema_by_name(column_name).is_none(),
-                error::ColumnExistsSnafu {
-                    column_name,
-                    table_name,
-                }
-            );
-        }
+        let original_primary_key_indices: HashSet<&usize> =
+            self.primary_key_indices.iter().collect();
 
         // Collect names of columns to add for error message.
-        let mut column_names = Vec::with_capacity(requests.len());
-        let mut primary_key_indices = self.primary_key_indices.clone();
+        let SplitResult {
+            columns_at_first,
+            columns_location_map,
+            columns_at_last,
+            column_names,
+        } = self.split_requests_by_column_location(table_name, requests)?;
+        let mut primary_key_indices = Vec::with_capacity(self.primary_key_indices.len());
         let mut columns = Vec::with_capacity(table_schema.num_columns() + requests.len());
-        columns.extend_from_slice(table_schema.column_schemas());
-        // Append new columns to the end of column list.
-        for request in requests {
-            column_names.push(request.column_schema.name.clone());
+        // add new columns with FIRST, and in reverse order of requests.
+        columns_at_first.iter().rev().for_each(|request| {
             if request.is_key {
                 // If a key column is added, we also need to store its index in primary_key_indices.
                 primary_key_indices.push(columns.len());
             }
             columns.push(request.column_schema.clone());
+        });
+        // add existed columns in original order and handle new columns with AFTER.
+        let dummy_location = &Vec::new();
+        for (index, schema) in table_schema.column_schemas().iter().enumerate() {
+            if original_primary_key_indices.contains(&index) {
+                primary_key_indices.push(columns.len());
+            }
+            columns.push(schema.clone());
+            columns_location_map
+                .get(&schema.name)
+                .unwrap_or(dummy_location)
+                .iter()
+                .rev()
+                .for_each(|request| {
+                    if request.is_key {
+                        // If a key column is added, we also need to store its index in primary_key_indices.
+                        primary_key_indices.push(columns.len());
+                    }
+                    columns.push(request.column_schema.clone());
+                });
```

Comment:
Don't see the necessity of `dummy_location`, maybe just:

```suggestion
            if let Some(requests) = columns_location_map.get(&schema.name) {
                requests.iter().rev().for_each(|request| {
                    if request.is_key {
                        // If a key column is added, we also need to store its index in primary_key_indices.
                        primary_key_indices.push(columns.len());
                    }
                    columns.push(request.column_schema.clone());
                });
```